cvss40.js

```diff
@@ -0,0 +1,1031 @@
+   *
+   * The process involves the following steps:
+   * 1. Determine the maximal scoring difference (MSD) for each EQ by computing the difference between
+   *    the current MacroVector and the next lower MacroVector.
+   * 2. Retrieve the highest severity vectors for each EQ and compute the severity distances from the
+   *    to-be-scored vector.
+   * 3. Calculate the current severity distances for each EQ and determine the proportion of the distance.
+   * 4. Compute the mean of the proportional distances.
+   * 5. Subtract the mean distance from the score of the highest severity vector to obtain the final score.
+   *
+   * For detailed information on the CVSS v4.0 specification, refer to:
+   * https://www.first.org/cvss/v4.0/specification-document
+   *
+   * @returns {number} - The calculated CVSS v4.0 base score, rounded to one decimal place.
+   */
+  calculateScore() {
+
+      // Constants
+      const NO_IMPACT_METRICS = ["VC", "VI", "VA", "SC", "SI", "SA"];
+      const STEP = 0.1;
+
+      // Exception for no impact on system
+      if (NO_IMPACT_METRICS.every((metric) => this.getEffectiveMetricValue(metric) === "N")) {
+          return 0.0;
+      }
+
+      let value = CVSS40.SCORE_MAPPING[this.macroVectorResult]
+
+      // 1. For each of the EQs:
+      //   a. The maximal scoring difference is determined as the difference
+      //      between the current MacroVector and the lower MacroVector.
+      //     i. If there is no lower MacroVector the available distance is
+      //        set to NaN and then ignored in the further calculations.
+
+      // EQ values
+      const [eq1, eq2, eq3, eq4, eq5, eq6] = this.macroVectorResult.split('').map(Number);
+
+
+      // compute next lower macro, it can also not exist
+      const eq1_next_lower_macro = `${eq1 + 1}${eq2}${eq3}${eq4}${eq5}${eq6}`;
+      const eq2_next_lower_macro = `${eq1}${eq2 + 1}${eq3}${eq4}${eq5}${eq6}`;
+
+      let eq3eq6_next_lower_macro;
+      let eq3eq6_next_lower_macro_left;
+      let eq3eq6_next_lower_macro_right;
+
+      // eq3 and eq6 are related
+      if (eq3 == 1 && eq6 == 1) {
+          // 11 --> 21
+          eq3eq6_next_lower_macro = "".concat(eq1, eq2, eq3 + 1, eq4, eq5, eq6)
+      } else if (eq3 == 0 && eq6 == 1) {
+          // 01 --> 11
+          eq3eq6_next_lower_macro = "".concat(eq1, eq2, eq3 + 1, eq4, eq5, eq6)
+      } else if (eq3 == 1 && eq6 == 0) {
+          // 10 --> 11
+          eq3eq6_next_lower_macro = "".concat(eq1, eq2, eq3, eq4, eq5, eq6 + 1)
+      } else if (eq3 == 0 && eq6 == 0) {
+          // 00 --> 01
+          // 00 --> 10
+          eq3eq6_next_lower_macro_left = "".concat(eq1, eq2, eq3, eq4, eq5, eq6 + 1)
+          eq3eq6_next_lower_macro_right = "".concat(eq1, eq2, eq3 + 1, eq4, eq5, eq6)
+      } else {
+          // 21 --> 32 (do not exist)
+          eq3eq6_next_lower_macro = "".concat(eq1, eq2, eq3 + 1, eq4, eq5, eq6 + 1)
```

Comment:
Out of curiosity, why using `"".concat(...)` here instead of `${eq1}$...` everywhere else ?

Reply:
That was easier to write when I was writing-it, not much more to say